go/vt/worker/key_resolver.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2015, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package worker
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/youtube/vitess/go/sqltypes"
+
+	"github.com/youtube/vitess/go/vt/key"
+	"github.com/youtube/vitess/go/vt/mysqlctl/tmutils"
+	tabletmanagerdatapb "github.com/youtube/vitess/go/vt/proto/tabletmanagerdata"
+	topodatapb "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/topo"
+)
+
+// This file defines the interface and implementations of sharding key resolvers.
+
+// keyspaceIDResolver defines the interface that needs to be satisifed to get a
+// keyspace ID from a database row.
+type keyspaceIDResolver interface {
+	// keyspaceID takes a table row, and returns the keyspace id as bytes.
+	// It will return an error if no sharding key can be found.
+	keyspaceID(row []sqltypes.Value) ([]byte, error)
+}
+
+// v2Resolver is the keyspace id resolver that is used by VTGate V2 deployments.
+// In V2, the sharding key column name and type is the same for all tables,
+// and the keyspace ID is stored in the sharding key database column.
+type v2Resolver struct {
+	keyspaceInfo        *topo.KeyspaceInfo
+	shardingColumnIndex int
+}
+
+// newV2Resolver returns a keyspaceIDResolver for a v2 table.
+// V2 keyspaces have a preset sharding column name and type.
+func newV2Resolver(keyspaceInfo *topo.KeyspaceInfo, td *tabletmanagerdatapb.TableDefinition) (keyspaceIDResolver, error) {
+	if keyspaceInfo.ShardingColumnName == "" {
+		return nil, errors.New("ShardingColumnName needs to be set for a v2 sharding key")
+	}
+	if keyspaceInfo.ShardingColumnType == topodatapb.KeyspaceIdType_UNSET {
+		return nil, errors.New("ShardingColumnType needs to be set for a v2 sharding key")
+	}
+	// Find the sharding key column index.
+	columnIndex := -1
+	if td.Type == tmutils.TableBaseTable {
```

Comment:
I think it would be easier to:
if td.Type != tmutils.TableBaseTable {
    return nil, errors.new(...)
}
in this case? it cannot happen anyway. Panic might work too.